### PR TITLE
Tests for Group Division

### LINKED ISSUE #24: Create Unit tests
N/A

### PR DESCRIPTION
**Pull request by TLE931**
written in PHP Pest framework. CI runs new tests automatically

### CODE REVIEW

File: tests/Feature/GroupCourseDivisionTest.php [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+use App\Models\Registration;
+use App\Models\Group;
+use App\Helpers\GroupCourseDivision;
+
+
+uses(\Illuminate\Foundation\Testing\RefreshDatabase::class);
+
+beforeEach(function () {
+    $this->event = Event::factory()->create(['consider_alcohol' => true])->first();
+    $this->course = Course::factory()->count(1)->create()->first();
```
Comment (simonostendorf) [inline tests/Feature/GroupCourseDivisionTest.php:15]: I think the course devision should be tested with more than one course.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -62,8 +62,6 @@ jobs:
...
         run: |
           php artisan db:seed
-      # - name: Execute tests (Unit and Feature tests) via PHPUnit
-      #   env:
-      #     DB_CONNECTION: sqlite
-      #     DB_DATABASE: database/database.sqlite
-      #   run: vendor/bin/phpunit
+      - name: Execute tests (Unit and Feature tests) via Pest
+        run: |
+          ./vendor/bin/pest
```
Comment (simonostendorf) [inline .github/workflows/ci.yml:67]: Is calling `./vendor/bin/pest` or `php artisan test` the correct/recommended way to execute the tests?

File: README.md [hunk 3]
```diff
@@ -113,7 +113,7 @@ npm run lint:fix
 You can run the tests with:
 
 ```sh
-vendor/bin/phpunit
+./vendor/bin/pest
```
Comment (simonostendorf) [inline README.md:116]: Is calling `./vendor/bin/pest` or `php artisan test` the correct/recommended way to execute the tests?

File: app/Helpers/GroupBalancedDivision.php
Comment (simonostendorf) [inline app/Helpers/GroupBalancedDivision.php:1]: The diff doesnt show the changes in a very good way, so I am asking this as a review comment: 

did you only reorder some things in the division or did you change the algorithm? If you changed the algorithm what did you change and does this fix #503 ?

File: tests/Unit/ExampleTest.php
Comment (simonostendorf) [inline tests/Unit/ExampleTest.php:1]: I think the example test should still exist to show the developers / users that there is a difference in feature and unit tests and that unit tests are possible.

**Review by simonostendorf — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 2 above)
Comment (TitusKirch) [reply .github/workflows/ci.yml:67]: `./vendor/bin/pest` is here the "best" way. `php artisan test` is for development and debugging  (see https://laravel.com/docs/11.x/testing#running-tests)

File: database/factories/CourseFactory.php [hunk 4]
```diff
@@ -19,7 +19,7 @@ public function definition(): array
         return [
             'name' => $this->faker->word(),
             'abbreviation' => $this->faker->lexify('????'),
-            'color' => $this->faker->word(),
+            'classes' => "bg-blue-700 text-white dark:bg-blue-800",
```
Comment (TitusKirch) [inline database/factories/CourseFactory.php:22]: It would be better to use a random element from an array here. Faker also has a direct method for this, which should look something like this (untested):

```PHP
'classes' => $this->faker->randomElement([
    'bg-yellow-500 text-white dark:bg-yellow-600',
    'bg-blue-700 text-white dark:bg-blue-800'
    // ... can be copied from the CourseSeeder
]),
```

File: database/factories/GroupFactory.php [hunk 5]
```diff
@@ -21,7 +21,6 @@ public function definition(): array
         return [
             'name' => $this->faker->word(),
             'event_id' => Event::all(['id'])->random(),
-            'course_id' => Course::all(['id'])->random(),
```
Comment (TitusKirch) [inline database/factories/GroupFactory.php:24]: Why was this removed?

File: phpunit.xml [hunk 6]
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.4/phpunit.xsd" bootstrap="vendor/autoload.php" colors="true" cacheDirectory=".phpunit.cache">
+  <testsuites>
+    <!--<testsuite name="Unit">
```
Comment (TitusKirch) [inline phpunit.xml:4]: If you remove the UnitTest, this should also be removed. But as mentioned above, I'll also write something about this in the UnitTest, so it's best to read that first.

File: tests/Feature/ExampleTest.php
Comment (TitusKirch) [inline tests/Feature/ExampleTest.php:1]: In my opinion, there should continue to be a feature test with this content, not as an example but so that you can continue to test via CI whether the app starts and runs. This is currently no longer being tested.

File: tests/Pest.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+| Test Case
+|--------------------------------------------------------------------------
+|
+| The closure you provide to your test functions is always bound to a specific PHPUnit test
+| case class. By default, that class is "PHPUnit\Framework\TestCase". Of course, you may
+| need to change it using the "uses()" function to bind a different classes or traits.
+|
+*/
+
+uses(Illuminate\Foundation\Testing\TestCase::class)->in('Feature', 'Unit');
```
Comment (TitusKirch) [inline tests/Pest.php:14]: There is also talk of unit tests here again, even though they have been removed. But see the comment below.

File: tests/Pest.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+| Expectations
+|--------------------------------------------------------------------------
+|
+| When you're writing tests, you often need to check that values meet certain conditions. The
+| "expect()" function gives you access to a set of "expectations" methods that you can use
+| to assert different things. Of course, you may extend the Expectation API at any time.
+|
+*/
+
+expect()->extend('toBeOne', function () {
```
Comment (TitusKirch) [inline tests/Pest.php:27]: The example/tutorial code can be removed

File: tests/Pest.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+| Functions
+|--------------------------------------------------------------------------
+|
+| While Pest is very powerful out-of-the-box, you may have some testing code specific to your
+| project that you don't want to repeat in every file. Here you can also expose helpers as
+| global functions to help you to reduce the number of lines of code in your test files.
+|
+*/
+
+function something()
```
Comment (TitusKirch) [inline tests/Pest.php:42]: The example/tutorial code can be removed

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+use App\Models\User;
+use App\Models\Course;
+use App\Models\Registration;
+use App\Models\Group;
+
+
+uses(\Illuminate\Foundation\Testing\RefreshDatabase::class);
+
+test('equal div only drinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:14]: Code is repeated redunantly here. Use the hooks from Pest such as `beforeEach`. See also: https://pestphp.com/docs/hooks

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+            expect($group->registrations()
+                ->join('users', 'registrations.user_id', '=', 'users.id')
+                ->where('users.course_id', '=', $course->id)
+                ->count())->toEqual(2);
+        }
+    }
+});
+
+test('unequal div only drinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:47]: Code is repeated redunantly here. Use the hooks from Pest such as `beforeEach`. See also: https://pestphp.com/docs/hooks

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+                ->join('users', 'registrations.user_id', '=', 'users.id')
+                ->where('users.course_id', '=', $course->id)
+                ->count())->toEqual($i);
+            $i += 1;
+        }
+    }
+});
+
+test('unequal div nondrinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:84]: Code is repeated redunantly here. Use the hooks from Pest such as `beforeEach`. See also: https://pestphp.com/docs/hooks

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    $courses = Course::factory()->count(10)->create();
+    $groups = Group::factory()->count(10)->create(['event_id' => $event->id]);
+
+    foreach ($courses as $course) {
+        User::factory()->count(20)->create(['course_id' => $course->id]);
+    }
+
+    foreach (User::all() as $user) {
+        Registration::factory()->create([
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:23]: All attributes are set here, what added value does it offer to use a factory here instead of making a direct assignment? In my opinion, this should be implemented here without a factory and using direct assignment.

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+
+    $i = 10;
+
+    foreach ($courses as $course) {
+        User::factory()->count($i)->create(['course_id' => $course->id]);
+        $i += 10;
+    }
+
+    foreach (User::all() as $user) {
+        Registration::factory()->create([
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:59]: All attributes are set here, what added value does it offer to use a factory here instead of making a direct assignment? In my opinion, this should be implemented here without a factory and using direct assignment.

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+uses(\Illuminate\Foundation\Testing\RefreshDatabase::class);
+
+test('equal div only drinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
+    $courses = Course::factory()->count(10)->create();
+    $groups = Group::factory()->count(10)->create(['event_id' => $event->id]);
+
+    foreach ($courses as $course) {
+        User::factory()->count(20)->create(['course_id' => $course->id]);
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:19]: I doubt whether this is a valid test method, in which you create the same number of users for each couirse in chronological order. It would be better to create e.g. 100 users with random courses and then see whether this is balanced. This can be done by chance so that they are simply assigned one after the other and are thus balanced.

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    }
+});
+
+test('unequal div only drinkers', function () {
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
+    $courses = Course::factory()->count(10)->create();
+    $groups = Group::factory()->count(10)->create(['event_id' => $event->id]);
+
+    $i = 10;
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:51]: An explanation for the $i would be useful (or a unique name). Otherwise it is difficult to understand at first glance.

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        foreach ($courses as $course) {
+            expect($group->registrations()
+                ->join('users', 'registrations.user_id', '=', 'users.id')
+                ->where('users.course_id', '=', $course->id)
+                ->count())->toEqual(2);
+        }
+    }
+});
+
+test('unequal div only drinkers', function () {
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:46]: The only thing that is uequal here in the test is that there are always 10 more users per degree programme. They are created again in chronological order. What added value does this test offer compared to the previous one?

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+            expect($group->registrations()
+                ->join('users', 'registrations.user_id', '=', 'users.id')
+                ->where('users.course_id', '=', $course->id)
+                ->count())->toEqual($i);
+            $i += 1;
+        }
+    }
+});
+
+test('unequal div nondrinkers', function () {
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:83]: The test should rather be called "drinkers and nondrinkers"

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+    $event = Event::factory()->create(['consider_alcohol' => true])->first();
+    $courses = Course::factory()->count(10)->create();
+    $groups = Group::factory()->count(10)->create(['event_id' => $event->id]);
+
+    $i = 10;
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:88]: Same thing as above

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+
+    foreach ($courses as $course) {
+        User::factory()->count($i)->create(['course_id' => $course->id]);
+        $i += 10;
+    }
+
+    $checks = 0;
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:95]: Commentary would be helpful

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+    $balance = 0.0;
+
+    for ($n = 0; $n < 10; $n++) {
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:98]: What added value is expected from this implementation? The use of Magic Numbers in particular should not occur at an arbitrary point in the test. Instead, the whole thing should be structured in such a way that the entire case is executed several times.

File: tests/Feature/GroupBalancedDivisionTest.php [hunk 22]
```diff
@@ -0,0 +1,148 @@
...
+                $checks++;
+            }
+        }
+
+        Registration::truncate();
+    }
+
+    // check average course balance (with allowed deviation due to priority for assigning non drinkers)
+    expect($balance / $checks)->toBeLessThanOrEqual(0.5);
+});
```
Comment (TitusKirch) [inline tests/Feature/GroupBalancedDivisionTest.php:148]: Due to the randomised system, however, edge cases are not taken into account with a 50/50 chance of someone drinking or not, if users are set up chronologically by study programme. Such as the existing bug here: https://github.com/fsr5-fhaachen/portals/issues/503

File: tests/Feature/SlotAssignmentTest.php [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+    $registration_to->modify('+7 days');
+
+    $this->event->registration_from = $registration_from;
+    $this->event->registration_to = $registration_to;
+    $this->event->has_requirements = false;
+    $this->event->consider_alcohol = false;
+    $this->event->sort_order = 300;
+    $this->event->save();
+
+    // cerate a new slot
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:30]: Typo: `create`

File: tests/Feature/SlotAssignmentTest.php [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+
+    $this->assignment = new SlotAssignment($this->slot);
+});
+
+function createUsers(int $amount)
+{
+    return User::factory()
+        ->has(
+            Registration::factory()
+                ->state([
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:48]: Why was it decided here to use the `state()` method for an inline state transformation?

File: tests/Feature/SlotAssignmentTest.php [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+                ])
+                ->count(1)
+        )
+        ->state(['course_id' => Course::first()->id])
+        ->count($amount)
+        ->create();
+}
+
+test('assign with more registrations than free slots', function () {
+    $reg_amount = 50;
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:65]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+
+    createUsers($reg_amount);
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:67]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+
+    // queue position should be equal to -1 before assignment                 
+    expect($this->slot->registrations()->where('queue_position', -1)->get()->count())->toEqual($reg_amount);
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:70]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+        if ($i < 10) {
+            expect($registrations[$i]->queue_position)->toBeNull();
+            continue;
+        }
+        expect($registrations[$i]->queue_position)->toEqual($i - 9);
+    }
+});
+
+test('assign with less registrations than slots', function () {
+    $reg_amount = 5;
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:88]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 29]
```diff
@@ -0,0 +1,136 @@
...
+
+    createUsers($reg_amount);
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:90]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+
+    $this->assignment->assign();
+
+    expect($this->slot->registrations()->where('queue_position', -1)->get()->count())->toEqual(0);
+
+    // queue position should be null for all registrations
+    expect($this->slot->registrations()->where('queue_position', null)->get()->count())->toEqual($reg_amount);
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:97]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+});
+
+test('assign with exact registrations', function () {
+    $reg_amount = 10;
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:101]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 32]
```diff
@@ -0,0 +1,136 @@
...
+
+    createUsers($reg_amount);
+
+    $this->assignment->assign();
+
+    expect($this->slot->registrations()->where('queue_position', -1)->get()->count())->toEqual(0);
+
+    // queue position should be null for all registrations
+    expect($this->slot->registrations()->where('queue_position', null)->get()->count())->toEqual($reg_amount);
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:110]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 33]
```diff
@@ -0,0 +1,136 @@
...
+});
+
+test('reassign after new registrations', function () {
+    $reg_amount = 5;
+    $reg_amount_later = 12;
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:115]: This should be in camelCase, as this is what we primarily use.

File: tests/Feature/SlotAssignmentTest.php [hunk 34]
```diff
@@ -0,0 +1,136 @@
...
+
+    createUsers($reg_amount);
```
Comment (TitusKirch) [inline tests/Feature/SlotAssignmentTest.php:117]: This should be in camelCase, as this is what we primarily use.

**Review by TitusKirch — CHANGES_REQUESTED** [review]

File: database/factories/GroupFactory.php (hunk 5 above)
Comment (simonostendorf) [reply database/factories/GroupFactory.php:24]: There is no course_id anymore, see PR #673

File: app/Helpers/GroupBalancedDivision.php
Comment (TLE931) [reply app/Helpers/GroupBalancedDivision.php:1]: if no non drinkers were registered, the method would run into an exception. Titus saw correctly, the if-query is the only change.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (TLE931) [reply .github/workflows/ci.yml:67]: I tried both and didn't see a difference

File: tests/Feature/SlotAssignmentTest.php (hunk 24 above)
Comment (TLE931) [reply tests/Feature/SlotAssignmentTest.php:48]: @jpih has to answer that

File: app/Helpers/GroupBalancedDivision.php
Comment (simonostendorf) [reply app/Helpers/GroupBalancedDivision.php:1]: Does the algorithm now work with 100% drinkers and 100% non drinkers?

File: tests/Feature/GroupCourseDivisionTest.php (hunk 1 above)
Comment (TLE931) [reply tests/Feature/GroupCourseDivisionTest.php]: outdated, CourseDivisionTest was rewritten

File: app/Helpers/GroupBalancedDivision.php
Comment (TLE931) [reply app/Helpers/GroupBalancedDivision.php:1]: it does not run into an error in these cases

File: tests/Unit/ExampleTest.php
Comment (TLE931) [reply tests/Unit/ExampleTest.php:1]: now contains ExampleTest for Unit and Feature

File: .github/workflows/ci.yml [hunk 35]
```diff
@@ -62,8 +62,6 @@ jobs:
           DB_DATABASE: database/database.sqlite
         run: |
           php artisan db:seed
-      # - name: Execute tests (Unit and Feature tests) via PHPUnit
-      #   env:
-      #     DB_CONNECTION: sqlite
-      #     DB_DATABASE: database/database.sqlite
-      #   run: vendor/bin/phpunit
+      - name: Execute tests (Unit and Feature tests) via Pest
```
[the remark this answers is not in the document]
Comment (TLE931) [reply .github/workflows/ci.yml:65]: now has unit and feature tests

File: database/factories/CourseFactory.php (hunk 4 above)
Comment (TLE931) [reply database/factories/CourseFactory.php]: done

File: tests/Feature/GroupBalancedDivisionTest.php (hunk 10 above)
Comment (TLE931) [reply tests/Feature/GroupBalancedDivisionTest.php]: done

File: tests/Feature/GroupBalancedDivisionTest.php (hunk 11 above)
Comment (TLE931) [reply tests/Feature/GroupBalancedDivisionTest.php]: done

File: tests/Feature/GroupBalancedDivisionTest.php (hunk 12 above)
Comment (TLE931) [reply tests/Feature/GroupBalancedDivisionTest.php]: done

File: tests/Feature/GroupBalancedDivisionTest.php (hunk 14 above)
Comment (TLE931) [reply tests/Feature/GroupBalancedDivisionTest.php:106]: done

File: tests/Feature/GroupBalancedDivisionTest.php (hunk 16 above)
Comment (TLE931) [reply tests/Feature/GroupBalancedDivisionTest.php]: new test contains explanations

File: tests/Feature/GroupBalancedDivisionTest.php (hunk 17 above)
Comment (TLE931) [reply tests/Feature/GroupBalancedDivisionTest.php]: rewrote test

File: tests/Feature/GroupBalancedDivisionTest.php (hunk 18 above)
Comment (TLE931) [reply tests/Feature/GroupBalancedDivisionTest.php]: done
